gdb_backend/program.c

```diff
@@ -1,6 +1,270 @@
+	remove_val_d(dlist, 12);
+
+    struct list_2 *list2 = malloc(sizeof(struct list_2));
+    list2->head = NULL;
+    appendList(5, list2);
+    appendList(10, list2);
+    appendList(21, list2);
+    appendList(234, list2);
+    appendList(67, list2);
+    print_list(list2);
+}
+
+void func1(int *arr, int len, int modify) {
+	for (int i = 0; i < len; i++) {
+		if (i % 2 == 0) {
+			arr[i] += modify;
+		} else {
+			arr[i] *= modify;
+		}
+	}
+}
+
+
+double slow_pow(double base, int exponent) {
+	int res = 1;
+	for (int i = 0; i < exponent; i++) {
+		res *= base;
+	}
+
+	return res;
+}
+
+double fast_pow(double base, int exponent) {
+	if (exponent == 0) {
+		return 1;
+	} else if (exponent == 1) {
+		return base;
+	} else {
+		if (exponent % 2 == 0) {
+			double temp = fast_pow(base, exponent/2);
+			return temp * temp;
+		} else {
+			double temp = fast_pow(base, exponent/2);
+			return temp * temp * base;
+		}
+	}
+}
+
+// Basic list functions
+#pragma region basic list functions
+
+struct node *new_node(int val) {
+	struct node *new = malloc(sizeof(struct node));
+	new->data = val;
+	new->next = NULL;
+	return new;
+}
+
+struct node *new_list(int val) {
+	return new_node(val);
+}
+
+void append(struct node * list, int val) {
+	struct node *new = new_node(val);
+	struct node *curr = list;
+	while (curr->next != NULL) {
+		curr = curr->next;
+	}
+	curr->next = new;
+}
+
+void remove_back(struct node * list) {
+	struct node *curr = list;
+	while (curr->next->next != NULL) {
+		curr = curr->next;
+	}
+	free(curr->next);
+	curr->next = NULL;
+}
+
+void remove_val(struct node * list, int val) {
+	struct node *curr = list;
+	while (curr->next != NULL) {
+		if (curr->next->data == val) {
+			struct node *temp = curr->next->next;
+			free(curr->next);
+			curr->next = temp;
+			break;
+		}
+		curr = curr->next;
+	}
+}
+
+struct node *remove_head(struct node * list) {
+	struct node *temp = list->next;
+	free(list);
+	return temp;
+}
+#pragma endregion
+
+#pragma region contained list functions
+
+noded_t *new_noded(double val) {
+	noded_t *new = malloc(sizeof(noded_t));
+	new->data = val;
+	new->next = NULL;
+	new->prev = NULL;
+	return new;
+}
+
+listd_t *new_listd(double val) {
+	noded_t *elem = new_noded(val);
+	listd_t *new = malloc(sizeof(listd_t));
+	new->head = elem;
+	new->tail = elem;
+	new->size = 0;
+	return new;
+}
+
+void appendd(listd_t *list, double val) {
+	noded_t *elem = new_noded(val);
+	elem->prev = list->tail;
+	list->tail->next = elem;
+	list->tail = elem;
+	list->size++;
+}
+
+void remove_backd(listd_t *list) {
+	noded_t *temp = list->tail;
+	list->tail = list->tail->prev;
+	list->tail->next = NULL;
+	free(temp);
+}
+
+void remove_val_d(listd_t *list, double val) {
+	noded_t *curr = list->head;
+	while (curr->next != NULL) {
+		if (curr->data == val) {
+			curr->prev->next = curr->next;
+			curr->next->prev = curr->prev;
+			free(curr);
+			break;
+		}
+		curr = curr->next;
+	}
+}
+
+void remove_headd(listd_t *list) {
```

Comment:
do we need to set the second node's prev to `NULL`?
e.g.
`list->head->next->prev = NULL`

otherwise it still points to the first node which will have been freed.

Reply:
That might be a good idea, but I was thinking, would it be useful to have intentionally buggy code like this, to test how the fe responds. Though it might be good to have both. What do you think?

Reply:
Oh yes definitely good to test how the FE will respond with buggy code. I think for testing it would be easiest if we wrote out both the proper version and the buggy version and have them both available. So name the buggy version `remove_headd_bugged()` or something. wdyt?

Reply:
At the very least we should put comments where the bug has been intentionally introduced

Reply:
I was thinking having different files with bugs in them, just be another event to change which one is being used. And yes we should document the bugs.

Reply:
Sure 👍

Reply:
oh hmm this wouldnt work for the cases where there are initially 0 or 1 nodes in the list